### PR TITLE
HasIO interface for IO actions

### PR DESCRIPTION
**Pull request by edwinb**
Also updates the Prelude and some base libraries to use HasIO rather than using IO directly. This is a draft, since it hasn't covered all the libraries yet (and some need a bit more thought, especially the concurrency primitives) and the documentation needs updating, especially FFI.

My main motivation is that whenever I experiment with some new abstraction for effects or linearity, I need to lift all the primitive IO operations somehow, and I wouldn't have to if they were in an appropriate interface. `HasIO` is basically like `MonadIO` from Haskell. My hope is that if everything lives in `HasIO` then we'll all get in the habit of adding primitives this way rather than putting them directly in `IO`!

* Is there a better name than `HasIO`? I'm not especially attached to it (but I don't like the name `MonadIO`)
* Are there other things that should get this treatment?
* Is there some potential problem I haven't thought of?

In case of worries about performance, I rebuilt Idris 2 using these libraries, then rebuilt again using the result, and the build times I'm getting for that result are within the usual range even though Idris2 itself does quite a lot of IO operations. So, at least, I'm not worried about that.

### CODE REVIEW

**Comment by edwinb** [comment]
Following some discussion on the slack, I've split this into `HasIO` and `MonadIO`. While I still don't especially like the name `MonadIO`, if it really does require `Monad` then I suppose it's fine...

File: docs/source/ffi/ffi.rst [hunk 1]
```diff
@@ -88,17 +88,19 @@ returns a primitive IO action:
...
 state of the world, and returns an ``Int`` with an updated state of the world.
-We can convert this into an ``IO`` action using ``primIO``:
+In general, ``IO`` operations in an Idris program are defined as instances
+of the ``HasIO`` interface (or ``MonadIO``). We can convert a primitive
+operation to one usable in ``HasIO`` using ``primIO``:
 
 .. code-block:: idris
 
-    primIO : PrimIO a -> IO a
+    primIO : HasIO a => PrimIO a -> io a
```
Comment (MarcelineVQ) [inline docs/source/ffi/ffi.rst:97]: Should say `HasIO io` iiuc

Comment (edwinb) [reply docs/source/ffi/ffi.rst:97]: oh my. How does this even work?

Comment (edwinb) [reply docs/source/ffi/ffi.rst:97]: Oh right, it's the documentation :). I'll fix it.

**Pull request merged**